Mathlib/Tactic/Linter/Lint.lean

```diff
@@ -269,4 +269,70 @@ initialize addLinter longLineLinter
 
 end LongLine
 
+/-! # The "openClassical" linter -/
+
+/-- The "openClassical" linter emits a warning on `open Classical` statements which are not
+scoped to a single declaration. A non-scoped `open Classical` can hide that some theorem statements
+would be better stated with explicit decidability statements.
+-/
+register_option linter.openClassical : Bool := {
+  defValue := true
+  descr := "enable the openClassical linter"
+}
+
+namespace OpenClassical
+
+/-- Gets the value of the `linter.openClassical` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.openClassical o
+
+/-- If `stx` is syntax describing an `open` command, `extractOpenNames stx`
+returns an array of for the opened names with their corresponding syntax
+(omitting any renamed or hidden items). -/
+def extractOpenNames : Syntax → Option (Array (Syntax × Name))
+  | `(command|open $openHiding:openHiding) =>
+    -- The first argument is the identifier we care about.
+    let arg := openHiding.raw[0]
+    some #[(arg, arg.getId)]
+  | `(command|open $openRenaming:openRenaming) =>
+    -- The first argument is the identifier we care about.
+    let arg := openRenaming.raw[0]
+    some #[(arg, arg.getId)]
+  | `(command|open $openOnly:openOnly) =>
+    -- The first argument is the identifier we care about.
+    some #[(openOnly.raw[0], openOnly.raw[0].getId)]
+  | `(command|open $openDecl:openSimple) =>
+    -- The first and only argument of `openDecl` is an array containing all the names we
+    -- are interested in.
+    let namesSyntax := ((openDecl.raw.getArgs).get! 0).getArgs
+    some (namesSyntax.map (fun arg ↦ (arg, arg.getId)))
+  | `(command|open $openScoped:openScoped) =>
+    -- The first argument of `openScoped` is "scoped",
+    -- the second one is an array containing all the names we are interested in.
+    let namesSyntax := (openScoped.raw[1]).getArgs
+    some (namesSyntax.map fun arg ↦ (arg, arg.getId))
+  | _ => none
+
+@[inherit_doc Mathlib.Linter.linter.openClassical]
+def openClassicalLinter : Linter where run := withSetOptionIn fun stx ↦ do
```

Comment:
```suggestion
def openClassicalLinter : Linter where run stx := do
```
Using a `set_option in` in front of a non-`in`-`set_option` seems like it would achieve to effective cancel the inner `set_option`.  Maybe this could be another linter, but here I don't think that `withSetOptionIn` is really needed.

Please double check, as I'm not completely sure!

Reply:
I just tried it out: making this change breaks the standard guard_msg dance: then, the following passes (i.e., the linter is silent)

set_option linter.openClassical false in
#guard_msgs in
set_option linter.openClassical true in
open Classical
```
Removing the `in` in the last line makes the linter complain, but the guard_msg empty - that's fun, but not really what I would expect.